### PR TITLE
Handle WSS shutdown and UDP buffer sizing

### LINKED ISSUE #97: [BUG] transport review findings: shutdown deadlock, nil-listener panic, websocket leak, and reconnect lifecycle issues
## Summary

I reviewed the current `master` branch of `AlexStocks/getty` locally and found several source-grounded lifecycle / concurrency / stability issues in `transport/` that look worth fixing.

The findings below are intentionally limited to issues with concrete code paths and clear impact, rather than broad speculation.

---

## 1. Client shutdown can deadlock because `client.stop()` closes sessions while holding `c.Lock()`, and session close can synchronously re-enter client reconnect logic

**Files**:
- `transport/client.go`
- `transport/session.go`

### Relevant code path

`client.stop()`:

```go
c.Lock()
for s := range c.ssMap {
    s.RemoveAttribute(sessionClientKey)
    s.RemoveAttribute(ignoreReconnectKey)
    s.Close()
}
c.Unlock()
```

`session.stop()`:

```go
clt, cltFound := s.GetAttribute(sessionClientKey).(*client)
ignoreReconnect, flagFound := s.GetAttribute(ignoreReconnectKey).(bool)
if cltFound && flagFound && !ignoreReconnect {
    clt.reConnect()
}
```

`reConnect()` calls methods that need the same client lock again (for example through `sessionNum()` / `connect()`).

### Why this matters

This creates a shutdown-time lock re-entry path:
- client shutdown holds `c.Lock()`
- session close runs synchronously under that lock
- session close can call back into `clt.reConnect()`
- reconnect path needs the same client lock again

That is a real deadlock risk during client shutdown.

### Suggested fix

- never call `s.Close()` while holding the client map lock
- first snapshot sessions under lock, then release the lock, then close them
- more generally, avoid synchronous reconnect work inside the session teardown path

---

## 2. Session teardown performs reconnect synchronously inside `session.stop()`, which makes close/shutdown paths unexpectedly heavy and fragile

**File**: `transport/session.go`

### Relevant code path

```go
if cltFound && flagFound && !ignoreReconnect {
    clt.reConnect()
}
```

`reConnect()` itself can:
- loop,
- dial,
- sleep/backoff,
- and inspect/update client session state.

### Why this matters

A session close path is expected to be teardown-only, but here it can synchronously perform reconnection logic.

That means:
- `Close()` is no longer a small/local operation,
- teardown can block inside reconnect backoff/dial logic,
- and shutdown ordering becomes harder to reason about.

This also amplifies the deadlock problem above.

### Suggested fix

- decouple reconnect scheduling from `session.stop()`
- move reconnect decisions to a dedicated client lifecycle controller / background worker
- make teardown publish state and return, rather than performing reconnection inline

---

## 3. Deferred error path in `handlePackage()` can panic because of an incorrect nil check

**File**: `transport/session.go`

### Relevant code path

```go
if err != nil {
    ...
    if s != nil || s.listener != nil {
        s.listener.OnError(s, err)
    }
}
```

The guard uses `||` instead of `&&`.

### Why this matters

If `s != nil` but `s.listener == nil`, the condition still passes and `s.listener.OnError(...)` dereferences a nil listener.

That means a normal I/O / decode / shutdown error in the deferred cleanup path can escalate into a panic.

### Suggested fix

- replace the guard with `if s != nil && s.listener != nil { ... }`
- add a regression test for the error-cleanup path with a nil listener

---

## 4. WebSocket self-connect rejection leaks the upgraded connection

**File**: `transport/server.go`

### Relevant code path

```go
conn, err := s.upgrader.Upgrade(w, r, nil)
...
if conn.RemoteAddr().String() == conn.LocalAddr().String() {
    log.Warnf(...)
    return
}
```

After the websocket upgrade succeeds, the self-connect branch returns directly without closing `conn`.

### Why this matters

At that point the connection has already been upgraded and allocated. Returning without closing it leaks the socket/resource for that request path.

### Suggested fix

- explicitly close `conn` before returning in the self-connect rejection branch
- add a test or at least a small harness covering this branch if possible

---

## 5. WSS event loop panics on normal server shutdown/error return paths

**File**: `transport/server.go`

### Relevant code path

```go
err = server.Serve(tls.NewListener(s.streamListener, config))
if err != nil {
    log.Errorf(...)
    panic(err)
}
```

The non-TLS WS loop logs serve errors and returns, but the WSS loop panics on any non-nil `Serve(...)` result.

### Why this matters

For long-running servers, `Serve(...)` returning is not always an exceptional condition. During normal shutdown, listener closure often causes `Serve` to return.

Treating that path as unconditional panic makes graceful shutdown brittle and can convert expected shutdown into process crash.

### Suggested fix

- distinguish expected shutdown errors from unexpected serve failures
- handle normal listener/server close without panic
- keep panic only for genuinely impossible internal invariants, if any

---

## 6. UDP receive buffer sizing contains a dead branch / likely logic bug

**File**: `transport/session.go`

### Relevant code path

```go
maxBufLen = int(s.maxMsgLen + maxReadBufLen)
if int(s.maxMsgLen<<1) < bufLen {
    maxBufLen = int(s.maxMsgLen << 1)
}
```

At this point `bufLen` has not yet been populated from `conn.recv(buf)` and is still zero.

### Why this matters

The condition is effectively dead on entry and does not do what it appears to intend.

This looks like a real logic bug in UDP buffer sizing, and it likely means the branch never contributes to runtime behavior.

### Suggested fix

- re-check the intended sizing rule
- if the intent was to cap by `2 * maxMsgLen`, compare against the right value instead of uninitialized `bufLen`
- add a focused test around UDP buffer sizing / message length guard behavior

---

## 7. Reconnect attempt accounting appears monotonic across successes, which may permanently disable reconnect after enough intermittent failures

**File**: `transport/client.go`

### Relevant code path

```go
var reconnectAttempts int
...
for {
    ...
    if connPoolSize <= sessionNum || maxReconnectAttempts < reconnectAttempts {
        break
    }
    c.connect()
    reconnectAttempts++
    ...
}
```

`reconnectAttempts` increases each loop, but I did not see it reset after a successful reconnect.

### Why this matters

If intermittent disconnect/reconnect cycles accumulate over process lifetime, the client may eventually stop reconnecting entirely once it crosses `maxReconnectAttempts`, even if previous reconnects actually succeeded.

### Suggested fix

- make retry accounting track consecutive reconnect failures rather than total lifetime attempts in the loop
- reset the counter when pool health is restored / a reconnect succeeds
- add a regression test for repeated disconnect-success-disconnect cycles

---

## Notes

I did **not** try to turn every suspicious pattern into an issue here. The items above are the ones that looked most actionable and had the clearest code evidence.

If helpful, I can also add a follow-up comment grouping these into:
- directly actionable fixes,
- and items that should first get a reproducer / stress test.

### PR DESCRIPTION
**Pull request by AlexStocks**
## Summary

- treat `http.ErrServerClosed` and Getty shutdown as normal WSS event-loop exits instead of panicking
- compute UDP receive-buffer capacity from configured message limits rather than the unread `bufLen` zero value
- add deterministic Issue #97 regression tests plus the approved design and implementation plan

## Scope

This PR addresses Issue #97 items 5 and 6 only. It intentionally does not close #97 because reconnect-attempt accounting remains a separate semantics decision.

## Validation

- [x] `go test -race ./transport -count=1`
- [x] `go vet ./...`
- [x] `go test ./... -count=1`
- [x] targeted regressions with `-race -count=20`
- [x] `gofmt` and `git diff --check`

## TDD evidence

- WSS regression failed before the fix with `panic: http: Server closed`
- UDP regression failed before the fix with `undefined: udpReadBufferSize`

## Summary by CodeRabbit

* **Bug Fixes**
  * WSS shutdowns now complete gracefully without panics or blocking.
  * Unexpected active-server errors remain visible in logs.
  * UDP receive buffering safely handles invalid, extreme, and large message sizes.
* **CI & Security**
  * Added automated testing, coverage, race detection, cross-platform builds, CodeQL analysis, and dependency updates.
  * Improved formatting and workflow reliability.
* **Tests**
  * Added regression and integration coverage for WSS shutdown and UDP buffer handling.
* **Documentation**
  * Updated CI status badges and runtime-fix documentation.

### CODE REVIEW

File: transport/session_test.go [hunk 1]
```diff
@@ -31,6 +31,29 @@ var (
...
+		{name: "tiny message", maxMsgLen: 1, want: 2},
+		{name: "below crossover", maxMsgLen: maxReadBufLen - 1, want: 2 * (maxReadBufLen - 1)},
+		{name: "at crossover", maxMsgLen: maxReadBufLen, want: 2 * maxReadBufLen},
+		{name: "above crossover", maxMsgLen: maxReadBufLen + 1, want: 2*maxReadBufLen + 1},
+		{name: "large message", maxMsgLen: 128 * 1024, want: 128*1024 + maxReadBufLen},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := udpReadBufferSize(tt.maxMsgLen); got != tt.want {
```
Comment (AlexStocks) [inline transport/session_test.go:55]: [P1] 回归测试没有覆盖生产调用链
当前测试只直接调用 `udpReadBufferSize`。我在 Head 副本中把 `handleUDPPackage` 的分配逻辑恢复为本 PR 修复前读取未赋值 `bufLen` 的分支，同时保留该 helper 和本测试；`go test ./transport -run '^TestUDPReadBufferSize$' -count=20` 仍然通过。因此生产路径即使重新引入 Issue #97 #6，这个测试也不会失败。建议补一个经 `handleUDPPackage`/真实 UDP session 触发分配规则的回归测试，或增加可观测的包内测试边界，并用恢复旧调用逻辑的变异确认测试会变红。

File: transport/session.go [hunk 2]
```diff
@@ -64,6 +64,14 @@ const (
 	outputFormat          = "session %s, Read Bytes: %d, Write Bytes: %d, Read Pkgs: %d, Write Pkgs: %d"
 )
 
+func udpReadBufferSize(maxMsgLen int32) int {
+	maxBufLen := int(maxMsgLen + maxReadBufLen)
+	if doubledMaxMsgLen := int(maxMsgLen << 1); doubledMaxMsgLen < maxBufLen {
+		return doubledMaxMsgLen
+	}
+	return maxBufLen
+}
```
[the remark this answers is not in the document]
Comment (AlexStocks) [reply transport/session.go:82]: 已在 1e2ea17 修复：非正 maxMsgLen 使用 64 KiB 有界回退，计算改为 int64 并限制到 UDP 数据报上限；SetMaxMsgLen 同时规范化负数和超出 int32 的输入。边界、setter、race 和 386 测试均通过。

File: transport/session_test.go [hunk 3]
```diff
@@ -31,6 +31,29 @@ var (
...
+		name      string
+		maxMsgLen int32
+		want      int
+	}{
+		{name: "tiny message", maxMsgLen: 1, want: 2},
+		{name: "below crossover", maxMsgLen: maxReadBufLen - 1, want: 2 * (maxReadBufLen - 1)},
+		{name: "at crossover", maxMsgLen: maxReadBufLen, want: 2 * maxReadBufLen},
+		{name: "above crossover", maxMsgLen: maxReadBufLen + 1, want: 2*maxReadBufLen + 1},
+		{name: "large message", maxMsgLen: 128 * 1024, want: 128*1024 + maxReadBufLen},
+	}
```
[the remark this answers is not in the document]
Comment (AlexStocks) [reply transport/session_test.go:51]: 已在 36680e8 增加 maxMsgLen=0、负值和 math.MaxInt32 边界，并补充 SetMaxMsgLen 规范化测试；修复前这些用例按预期失败，修复后通过。

File: transport/session_test.go (hunk 1 above)
Comment (AlexStocks) [reply transport/session_test.go:55]: 已在 36680e8 增加真实 net.UDPConn/newUDPSession/handleUDPPackage 调用链测试。临时恢复旧 bufLen 分配逻辑后，测试观察到 ReadFromUDP(p:4097) 且 Reader 收到 3 字节而不是 2 字节，按预期失败；恢复 helper 调用后通过，目标 race 测试重复 20 次通过。
